File: python/paddle_edl/discovery/etcd_client.py
# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import etcd3 as etcd
import etcd3.exceptions as exceptions
import functools
import json
import random
import logging
import time


class NoValidEndpoint(Exception):
    pass


class ServerMeta(object):
    def __init__(self, server, info, mod_revision, revision):
        self.server = server
        self.info = info
        self.mod_revision = mod_revision  # key mod revision
        self.revision = revision  # etcd global revision

    def __str__(self):
        return 'server={}, info={}, mod_revision={}, revision={}'.\
            format(self.server, self.info, self.mod_revision, self.revision)


def _handle_errors(f):
    def handler(*args, **kwargs):
        try:
            return f(*args, **kwargs)
        except exceptions.Etcd3Exception as e:
            args[0]._connect(args[0]._endpoints)

        return f(*args, **kwargs)

    return functools.wraps(f)(handler)


class EtcdClient(object):
    def __init__(self,
                 endpoints=['127.0.0.1:2379'],
                 passwd=None,
                 root="service"):
        self._endpoints = set(endpoints)
        self._passwd = passwd
        self._etcd = None
        self._leases = {}
        self._root = root

    def _endpoint_to_ip_port(self, endpoint):
        a = endpoint.split(":")
        return a[0], a[1]

    def _connect(self, endpoints):
        conn = None
        ep_lst = list(endpoints)
        random.shuffle(ep_lst)
        for ep in ep_lst:
            try:
                ip, port = self._endpoint_to_ip_port(ep)
                conn = etcd.client(host=ip, port=port)
            except Exception as e:
                print(e)
                continue

            self._etcd = conn
            return

        if self._etcd == None:
            raise NoValidEndpoint()

    def init(self):
        return self._connect(self._endpoints)

    @_handle_errors
    def get_service(self, service_name):
        servers = []
        d = '/{}/{}/nodes/'.format(self._root, service_name)
        for value, meta in self._etcd.get_prefix(d):
            servers.append(
                ServerMeta(
                    self.get_server_name_from_full_path(
                        meta.key, service_name), value, meta.mod_revision,
                    meta.response_header.revision))
        return servers

    @_handle_errors
    def get_service_with_revision(self, service_name):
        servers = []
        d = '/{}/{}/nodes/'.format(self._root, service_name)

        key_prefix = d
        range_response = self._etcd.get_prefix_response(key_prefix)
        for kv in range_response.kvs:
            servers.append(
                ServerMeta(
                    self.get_server_name_from_full_path(kv.key, service_name),
                    kv.value, kv.mod_revision, range_response.header.revision))

        return servers, range_response.header.revision

    def watch_service(self, service_name, call_back, **kwargs):
        # call_back(add_servers, rm_servers)
        #   add_servers: list(ServerMeta)
        #   rm_servers: list(ServerMeta)
        # start_revision=, watch start from revision
        def services_change(response):
            add_servers = dict()
            rm_servers = dict()
            for event in response.events:
                key = self.get_server_name_from_full_path(event.key,
                                                          service_name)
                server = ServerMeta(key, event.value, event.mod_revision,
                                    response.header.revision)

                if isinstance(event, etcd.events.PutEvent):
                    if key in rm_servers:
                        # after rm key, add again, need remove key from rm set
                        rm_servers.pop(key)
                    add_servers[key] = server
                elif isinstance(event, etcd.events.DeleteEvent):
                    if key in add_servers:
                        # after add key, rm again, need remove key from add dict
                        add_servers.pop(key)
                    rm_servers[key] = server
                else:
                    raise TypeError('ectd event type is not put or delete!')

            if len(add_servers) == 0 and len(rm_servers) == 0:
                return

            call_back(add_servers.values(), rm_servers.values())

        d = '/{}/{}/'.format(self._root, service_name)
        return self._etcd.add_watch_prefix_callback(d, services_change,
                                                    **kwargs)

    def cancel_watch(self, watch_id):
        return self._etcd.cancel_watch(watch_id)

    @_handle_errors
    def remove_service(self, service_name):
        d = '/{}/{}/'.format(self._root, service_name)
        servers = self.get_service(service_name)
        for s in servers:
            self.remove_server(service_name, s.server)
        self._etcd.delete_prefix(d)

    @_handle_errors
    def _get_lease(self, key, ttl=10):
        if key not in self._leases:
            lease = self._etcd.lease(ttl)
            self._leases[key] = lease

        return self._leases[key]

    @_handle_errors
    def set_server_not_exists(self,
                              service_name,
                              server,
                              info,
                              ttl=10,
                              timeout=20):
        """
        :returns: state of transaction, ``True`` if the put was successful,
                  ``False`` otherwise
        """
        key = '/{}/{}/nodes/{}'.format(self._root, service_name, server)
        lease = self._get_lease(key, ttl)
        begin = time.time()
        while True:
            if self._etcd.put_if_not_exists(key=key, value=info, lease=lease):
                return True

            # refresh lease
            for r in self._etcd.refresh_lease(lease.id):
                pass

            if (time.time() - begin > timeout):
                break

            time.sleep(1)
        return False

    @_handle_errors
    def _set_server(self, service_name, server, info, ttl=10):
        key = '/{}/{}/nodes/{}'.format(self._root, service_name, server)
        lease = self._get_lease(key, ttl)
        return self._etcd.put(key=key, value=info, lease=lease)

    @_handle_errors
    def _get_server(self, service_name, server):
        # for debug
        key = '/{}/{}/nodes/{}'.format(self._root, service_name, server)
        value, meta = self._etcd.get(key=key)
        return value, meta.key, meta.version, meta.create_revision, meta.mod_revision

    @_handle_errors
    def remove_server(self, service_name, server):
        key = '/{}/{}/nodes/{}'.format(self._root, service_name, server)
        self._etcd.delete(key)
        if key in self._leases:
            lease = self._leases[key]

            self._etcd.revoke_lease(lease.id)
            self._leases.pop(key)

    @_handle_errors
    def refresh(self, service_name, server, info=None, ttl=10):
        if info is not None:
            self._set_server(service_name, server, info, ttl)
            return

        key = '/{}/{}/nodes/{}'.format(self._root, service_name, server)
        lease = self._get_lease(key, ttl)

        for r in self._etcd.refresh_lease(lease.id):
            pass
        return

    def get_server_name_from_full_path(self, path, service_name):
        d = '/{}/{}/nodes/'.format(self._root, service_name)
        return path[len(d):]

    @_handle_errors
    def lock(self, service_name, server, ttl=10):
        key = '/{}/{}/nodes/{}'.format(self._root, service_name, server)
        return self._etcd.lock(key, ttl=ttl)

    @_handle_errors
    def get_key(self, key):
        return self._etcd.get(key)

File: pkg/master/etcd_client.go
// Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved.

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

// http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package master

import (
	"context"
	//"encoding/json"
	"fmt"
	"time"

	log "github.com/inconshreveable/log15"
	"go.etcd.io/etcd/clientv3"
	"go.etcd.io/etcd/clientv3/concurrency"
)

const (
	// DefaultLockPath is the default etcd master lock path.
	DefaultLockPath = "/master/lock"
	// DefaultStatePath is the default etcd key for master state.
	DefaultStatePath = "/master/state"
	// DefaultMetaPath is the default etcd key for master address.
	// {old_stages:[], now_stage:"", endpoint:""}
	DefaultMetaPath = "/master/meta"
	// DefaultAdjustPath is the the default etcd key for pod changes.
	// operation1, operatrion2, ...
	DefaultAdjustPath = "/master/adjust_operation"
)

const (
	dialTimeout = 5 * time.Second
)

// EtcdClient is the etcd client that the master uses for fault
// tolerance and service registry.
type EtcdClient struct {
	client *clientv3.Client
	lock   *concurrency.Mutex
	sess   *concurrency.Session
	jobID  string
}

type meta struct {
	JobStage string `json:"job_stage"`
	Endpoint string `json:"endpoint"`
}

func lockPath(jobID string) string {
	return fmt.Sprintf("/%s%s", jobID, DefaultLockPath)
}

func metaPath(jobID string) string {
	return fmt.Sprintf("/%s%s", jobID, DefaultMetaPath)
}

func statePath(jobID string) string {
	return fmt.Sprintf("/%s%s", jobID, DefaultStatePath)
}

func (e *EtcdClient) detectIsOwner() {
	for {
		err := e.lock.Lock(context.Background())
		if err != nil {
			s := "Master miss lock now, so exit!"
			log.Info(s)
			panic(s)
		}

		time.Sleep(5 * time.Second)
	}
}

func (e *EtcdClient) recordMeta(addr string) error {
	/*
		m := meta{"INITIAL", addr}

		d, _ := json.Marshal(m)

		return e.Save(d, metaPath(jobID))
	*/
	return nil
}

func (e *EtcdClient) loadState() (*masterState, error) {
	/*
		state, err := e.Load(statePath(jobID))
		if err != nil {
			log.Crit("load state info error:", err)
			return nil, err
		}

		return state, nil
	*/
	return nil, nil
}

func (e *EtcdClient) loadMeta() (*masterMeta, error) {
	/*
		meta, err := e.Load(metaPath(e.jobID))
		if err != nil {
			log.Crit("load state info error:", err)
			return nil, err
		}

		return meta, nil
	*/
	return nil, nil
}

// Election gets the master.
func Election(jobID string, endpoints []string, ttlSec int) *EtcdClient {
	/*
		var cli *EtcdClient
		cli.jobID = jobID
		var err error
		for {
			cli, err = tryToLock(jobID, endpoints, ttlSec)
			if err != nil {
				log.Debug("Elect failed:", err)
				time.Sleep(1 * time.Minute)
				continue
			}

			go detectIsOwner(jobID, cli)
			return cli
		}
	*/
	return nil
}

func tryToLock(jobID string, endpoints []string, ttlSec int) (*EtcdClient, error) {
	/*
		log.Debug("Connecting to etcd", log.Ctx{"endpoint": endpoints})
		cli, err := clientv3.New(clientv3.Config{
			Endpoints:   endpoints,
			DialTimeout: dialTimeout,
		})
		if err != nil {
			return nil, err
		}

		sess, err := concurrency.NewSession(cli, concurrency.WithTTL(ttlSec))
		if err != nil {
			return nil, err
		}

		lock := concurrency.NewMutex(sess, lockPath(jobID))
		// It's fine for the lock to get stuck, in this case we have
		// multiple master servers running (only configured to have
		// one master running, but split-brain problem may cause
		// multiple master servers running), and the cluster management
		// software will kill one of them.
		log.Info("Trying to acquire lock.", log.Ctx{"path": lockPath(jobID)})
		err = lock.Lock(context.TODO())
		if err != nil {
			return nil, err
		}
		log.Info("Successfully acquired lock at: ", log.Ctx{"path": lockPath(jobID)})

		e := &EtcdClient{
			client: cli,
			lock:   lock,
			sess:   sess,
		}

		//statePath string
		return e, nil
	*/
	return nil, nil
}

func (e *EtcdClient) reGetLock(miss bool) {
	/*
		if miss {
			log.Error("No longer owns the lock, trying to lock again")
		}
		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		err := e.lock.Lock(ctx)
		cancel()
		if err != nil {
			// We lost the master lock and can not acquire
			// it back, it means some other master is
			// already started. We don't want cluster
			// management system to kill the master server
			// who is holding the lock and running
			// correctly. So the most feasible solution is
			// to kill current master server. The current
			// state is not saved, but the trainer's RPC
			// call will fail, so the trainer will retry.
			log.Crit("Could not acquire the lock at %s: %v. Exiting.", log.Ctx{"path": lockPath(jobID), "error": err})
			panic("Could not acquire the lock at %s: %v. Exiting.")
		}
		if miss {
			log.Info("Successfully acquired lock at %s.", lockPath(e.jobID))
		}
	*/
	return
}

// Save saves the state into the etcd.
func (e *EtcdClient) Save(path string, v []byte) error {
	ctx := context.TODO()
	put := clientv3.OpPut(path, string(v))
	resp, err := e.client.Txn(ctx).If(e.lock.IsOwner()).Then(put).Commit()
	if err != nil {
		return err
	}

	if !resp.Succeeded {
		e.reGetLock(true)
		return e.Save(path, v)
	}

	return nil
}

// Load load value from etcd by path.
func (e *EtcdClient) Load(path string) ([]byte, error) {
	ctx := context.TODO()
	get := clientv3.OpGet(path)

	resp, err := e.client.Txn(ctx).If(e.lock.IsOwner()).Then(get).Commit()
	if err != nil {
		return nil, err
	}

	if !resp.Succeeded {
		e.reGetLock(true)
		return e.Load(path)
	}

	kvs := resp.Responses[0].GetResponseRange().Kvs
	if len(kvs) == 0 {
		// No state exists
		return nil, nil
	}

	v := kvs[0].Value
	return v, nil
}

// Shutdown shuts down the etcd client gracefully.
func (e *EtcdClient) Shutdown() error {
	err := e.sess.Close()
	newErr := e.client.Close()
	if newErr != nil {
		if err == nil {
			err = newErr
		} else {
			log.Error("shutdown error", log.Ctx{"error": newErr})
		}
	}

	return err
}

// GetKey gets the value by the specify key.
func GetKey(c *clientv3.Client, key string, timeout time.Duration) (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	resp, err := c.Get(ctx, key)
	cancel()
	if err != nil {
		return "", err
	}
	kvs := resp.Kvs
	if len(kvs) == 0 {
		return "", nil
	}
	v := kvs[0].Value
	return string(v), nil
}

// watchKey watches the specify key and send to valChan if there is some event.
func watchKey(c *clientv3.Client, key string, valChan chan<- string) {
	rch := c.Watch(context.Background(), key)
	for wresp := range rch {
		for _, ev := range wresp.Events {
			// if received event is DELETE, the value will be an empty string
			log.Info("received event.", log.Ctx{"type": ev.Type, "key": ev.Kv.Key, "value": ev.Kv.Value})
			valChan <- string(ev.Kv.Value)
		}
	}
}
